crates/noirc_evaluator/src/ssa/block.rs

```diff
@@ -298,35 +346,84 @@ pub fn rewire_block_left(ctx: &mut SsaContext, block_id: BlockId, left: BlockId)
     }
 }
 
+//Delete all instructions in the block
+pub fn short_circuit_block(ctx: &mut SsaContext, block_id: BlockId) {
+    let instructions = ctx[block_id].instructions.clone();
+    short_circuit_instructions(ctx, &instructions);
+}
+
+pub fn short_circuit_instructions(ctx: &mut SsaContext, instructions: &Vec<NodeId>) {
+    let mut zeros = HashMap::new();
+    for i in instructions {
+        let ins = ctx.get_instruction(*i);
+        if ins.res_type != node::ObjectType::NotAnObject {
+            zeros.insert(ins.res_type, ctx.zero_with_type(ins.res_type));
+        }
+    }
+    for i in instructions {
+        let ins = ctx.get_mut_instruction(*i);
+        if ins.res_type != node::ObjectType::NotAnObject {
+            ins.mark = Mark::ReplaceWith(zeros[&ins.res_type]);
+        } else if ins.operation.opcode() != Opcode::Nop {
+            ins.mark = Mark::Deleted;
+        }
+    }
+}
+
 //merge subgraph from start to end in one block, excluding end
-pub fn merge_path(ctx: &mut SsaContext, start: BlockId, end: BlockId) -> VecDeque<BlockId> {
+pub fn merge_path(
+    ctx: &mut SsaContext,
+    start: BlockId,
+    end: BlockId,
+    assumption: Option<NodeId>,
+) -> VecDeque<BlockId> {
     let mut removed_blocks = VecDeque::new();
     if start != end {
         let mut next = start;
         let mut instructions = Vec::new();
         let mut block = &ctx[start];
+        let mut short_circuit = BlockId::dummy();
+
         while next != end {
             if block.dominated.len() > 1 || block.right.is_some() {
                 dbg!(&block);
                 unreachable!("non sequential block sequence");
             }
             block = &ctx[next];
             removed_blocks.push_back(next);
-            instructions.extend(&block.instructions);
+
+            if short_circuit.is_dummy() {
+                instructions.extend(&block.instructions);
+            }
+
+            if short_circuit.is_dummy() && block.is_short_circuit(ctx, assumption) {
+                instructions.clear();
+                instructions.extend(&block.instructions);
```

Comment:
Do we need to clear the instructions vector and then extend it again here? Isn't this happening in the if block directly above where `if short_circuit.is_dummy()` is checked?